onnxruntime/core/providers/cuda/tensor/quantize_linear.cc

```diff
@@ -7,36 +7,148 @@
 namespace onnxruntime {
 namespace cuda {
 
+void ValidateBlockQuantizationShapes(const TensorShape& input_shape,
+                                     const TensorShape& scale_shape,
+                                     const Tensor* zero_point,
+                                     size_t axis_no_neg,
+                                     int64_t block_size_) {
+  ORT_ENFORCE(scale_shape.NumDimensions() == input_shape.NumDimensions(),
+              "scale and input must have the same rank for blocked quantization");
+
+  for (size_t i = 0, ndim = input_shape.NumDimensions(); i < ndim; ++i) {
+    if (i == static_cast<size_t>(axis_no_neg)) {
+      ORT_ENFORCE(scale_shape[i] == (input_shape[i] + block_size_ - 1) / block_size_,
+                  "scale must be ceil(Di/block_size) on the quantize axis i for blocked quantization");
+    } else {
+      ORT_ENFORCE(scale_shape[i] == input_shape[i],
+                  "scale and input must have the same shape despite the quantize axis for blocked quantization");
+    }
+  }
+
+  if (zero_point) {
+    ORT_ENFORCE(zero_point->Shape() == scale_shape,
+                "zero_point and scale must have the same shape for blocked quantization");
+  }
+}
+
 template <class T, class U>
-typename std::enable_if<boost::mp11::mp_set_contains<TypeList<int8_t, uint8_t>, T>::value, Status>::type
-CudaQuantizeLinear(cudaStream_t stream, const U* input, T* output, const U* scale, const T* zero_point, size_t num_of_element, bool /*saturate*/) {
+Status CudaQuantizeLinear(cudaStream_t stream, const U* input, T* output, const U* scale, const T* zero_point,
```

Comment:
>CudaQuantizeLinear [](http://example.com/codeflow?start=7&length=18)

what is the advantage to specialize it?

Reply:
It is for backward compatibility. opset 19-20 supports int8, fp32/fp16. opset 21 supports int4, fp32/fp16. However, the `ComputeInternal` is shared by all instantiations. That means int8 instantiation expects CudaQuantizeLinear, CudaQuantizeLinearAxis and CudaQuantizeLinearBlock to exist. The same for int4. If I keep the `std::enable_if<boost::mp11::mp_set_contains<TypeList<int8_t, uint8_t>, T>...`, the int8 instantiation will complain no instances of CudaQuantizeLinearBlock, and int4 will complain no instances of CudaQuantizeLinear or CudaQuantizeLinearAxis. So I have to add these placeholders for type compatibility.

Reply:
One way to avoid these placeholders is to add `std::enable_if` to `ComputeInternal` body, but that means `ComputeInternal` has the knowledge of the opset difference. I prefer to let the implementation itself handle the type difference.